node/libs/storage/src/block_store.rs

```diff
@@ -0,0 +1,218 @@
+                    .await?
+                    .cache
+                    .first_key_value()
+                    .unwrap()
+                    .1
+                    .clone();
+                self.0.persistent.store_next_block(ctx, &block).await?;
+                self.0.inner.send_modify(|inner| {
+                    debug_assert!(inner.persisted.next() == block.header().number);
+                    inner.persisted.last = block.justification.clone();
+                    inner.cache.remove(&block.header().number);
+                });
+            }
+        }
+        .await;
+        match res {
+            Ok(()) | Err(ctx::Error::Canceled(_)) => Ok(()),
+            Err(ctx::Error::Internal(err)) => Err(err),
+        }
+    }
+}
+
+impl BlockStore {
+    /// Constructs a BlockStore.
+    /// BlockStore takes ownership of the passed PersistentBlockStore,
+    /// i.e. caller should modify the underlying persistent storage
+    /// (add/remove blocks) ONLY through the constructed BlockStore.
+    pub async fn new(
+        ctx: &ctx::Ctx,
+        persistent: Box<dyn PersistentBlockStore>,
+        cache_capacity: usize,
+    ) -> ctx::Result<(Arc<Self>, BlockStoreRunner)> {
+        if cache_capacity < 1 {
+            return Err(anyhow::anyhow!("cache_capacity has to be >=1").into());
+        }
+        let state = persistent
+            .state(ctx)
+            .await?
+            .context("storage empty, expected at least 1 block")?;
+        if state.first.header().number > state.last.header().number {
+            return Err(anyhow::anyhow!("invalid state").into());
+        }
+        let this = Arc::new(Self {
+            persistent,
+            inner: sync::watch::channel(Inner {
+                inmem: sync::watch::channel(state.clone()).0,
+                persisted: state,
+                cache: BTreeMap::new(),
+                cache_capacity,
+            })
+            .0,
+        });
+        Ok((this.clone(), BlockStoreRunner(this)))
+    }
+
+    /// Fetches a block (from cache or persistent storage).
+    pub async fn block(
+        &self,
+        ctx: &ctx::Ctx,
+        number: validator::BlockNumber,
+    ) -> ctx::Result<Option<validator::FinalBlock>> {
+        {
+            let inner = self.inner.borrow();
+            if !inner.inmem.borrow().contains(number) {
+                return Ok(None);
+            }
+            if let Some(block) = inner.cache.get(&number) {
+                return Ok(Some(block.clone()));
+            }
+        }
+        Ok(Some(
+            self.persistent
+                .block(ctx, number)
+                .await?
+                .context("block disappeared from storage")?,
+        ))
+    }
+
+    /// Inserts a block to cache.
+    /// Since cache is a continuous range of blocks, `queue_block()`
+    /// synchronously waits until all intermediate blocks are added
+    /// to cache AND cache is not full.
```

Comment:
IMO, these constraints make it difficult to reason about store usage correctness; e.g., it's moderately easy to deadlock it by queuing blocks out of order. A somewhat artificial example: in `SyncBlocks` actor, if the number of concurrently downloaded blocks is more than cache capacity and the corresponding permits were (incorrectly) dropped too late, after queuing the received block.

Reply:
Can you clarify your example? As I understand it:
* N = download permits in SyncBlocks actor
* M = max cache size
* N > M
* first N blocks are being downloaded in parallel
* (what happens here?)

I can remove the block cache size by replacing queue_block with store_block (i.e. make fetching tasks until the block is stored persistently), which actually simplify the api.

Reply:
done